### DISCUSSION TITLE
Understanding multilayer output

### DISCUSSION THREAD

**Discussion by u-xixi** [mapequation/infomap#267] [Q&A] [answered]
hello Infomap devs,

I hope this is gonna be a quick question. I'm using Infomap with a multilayer network, and I just noticed that in the output, every node always ends up in its own module, beside being assigned to a big cluster. It looks like this in the ftree file:

> 6:14 0.00263657 "NODE_125" 125
> 125:1 0 "NODE_125" 125

I tested it by making a network of a single layer, which is a graph consists only of complete subgraphs that are totally isolated from each other. there are about 30 such subgraphs, I thought I'd see 30 modules, but infomap partition it into > 2000 top modules. I looked into it and see every node in their own module again. Is that normal? If so, why does it work that way and what's good about it? 
I was using the python API. infomap 2.0.0, mainly the three methods below 

```
im = Infomap("-s 1")
im.add_node()
im.add_multilayer_link()
```

thanks in advance
xixi

**Comment by danieledler** [comment, marked as answer]
Hi,

You are watching the physically aggregated tree output, where one physical node can exist in multiple modules. This is because each physical node may exist as multiple state nodes, one for each layer, and the state nodes may be part of different modules. If you check the state tree output, node 125 will be differentiated by different layer ids in the two different modules.

Do you add inter-layer links or only intra-layer links? If the latter, you can use [`add_multilayer_intra_link`](https://mapequation.github.io/infomap/python/infomap.html#infomap.Infomap.add_multilayer_intra_link) instead and let Infomap generate the links between layers, based on the `multilayer_relax_rate`.

Just to avoid possible sources of error, can you skip the calls to `add_node()` and see if you get the same result? The `add_multilayer_link()` method adds the state nodes as needed. If you want names in the output, use [`set_name`](https://mapequation.github.io/infomap/python/infomap.html#infomap.Infomap.set_name).

If you still get unexpected result, can you share more of your code and data?

**Reply by u-xixi** [reply]
thanks! I simply commented out add_nodes() and it works as expected now. I simply thought `add_node() `works just like `*Vertices x` field in the text input.
I'm working on a program using Infomap. It is a big network, and the physical nodes have a few attributes. so I made a layer of each attribute and hope Infomap can integrate them. The edges either stand for their similarity or their interaction. So i assume mine only has intra links, I just didn't see `add_multilayer_intra_link` before. In my code, I parse my user input and add links on the fly. Once it is done, I call `im.run()`.
